selfdrive/locationd/torqued.py

```diff
@@ -0,0 +1,267 @@
+#!/usr/bin/env python3
+import os
+import sys
+import signal
+import numpy as np
+from collections import deque, defaultdict
+
+import cereal.messaging as messaging
+from cereal import car, log
+from common.params import Params
+from common.realtime import config_realtime_process, DT_MDL
+from common.filter_simple import FirstOrderFilter
+from system.swaglog import cloudlog
+from selfdrive.controls.lib.vehicle_model import ACCELERATION_DUE_TO_GRAVITY
+
+
+HISTORY = 5  # secs
+POINTS_PER_BUCKET = 2000
+MIN_POINTS_TOTAL = 4000
+FIT_POINTS_TOTAL = 3000
+MIN_VEL = 15  # m/s
+FRICTION_FACTOR = 1.5  # ~85% of data coverage
+FACTOR_SANITY = 0.3
+FRICTION_SANITY = 0.5
+STEER_MIN_THRESHOLD = 0.02
+MIN_FILTER_DECAY = 50
+MAX_FILTER_DECAY = 250
+LAT_ACC_THRESHOLD = 1
+STEER_BUCKET_BOUNDS = [(-0.5, -0.3), (-0.3, -0.2), (-0.2, -0.1), (-0.1, 0), (0, 0.1), (0.1, 0.2), (0.2, 0.3), (0.3, 0.5)]
+MIN_BUCKET_POINTS = [100, 300, 500, 500, 500, 500, 300, 100]
+MAX_RESETS = 5.0
+MAX_INVALID_COUNT = 10
+MIN_ENGAGE_BUFFER = 2  # secs
+
+VERSION = 1  # bump this to invalidate old parameter caches
+
+
+def slope2rot(slope):
+  sin = np.sqrt(slope**2 / (slope**2 + 1))
+  cos = np.sqrt(1 / (slope**2 + 1))
+  return np.array([[cos, -sin], [sin, cos]])
+
+
+class PointBuckets:
+  def __init__(self, x_bounds, min_points):
+    self.x_bounds = x_bounds
+    self.buckets = {bounds: deque(maxlen=POINTS_PER_BUCKET) for bounds in x_bounds}
+    self.buckets_min_points = {bounds: min_point for bounds, min_point in zip(x_bounds, min_points)}
+
+  def bucket_lengths(self):
+    return [len(v) for v in self.buckets.values()]
+
+  def __len__(self):
+    return sum(self.bucket_lengths())
+
+  def is_valid(self):
+    return all([len(v) >= min_pts for v, min_pts in zip(self.buckets.values(), self.buckets_min_points.values())]) and (self.__len__() >= MIN_POINTS_TOTAL)
```

Comment:
no need for a list comprenhension with `any` and `all`. would also split this out into multiple lines